burr/core/state.py

```diff
@@ -2,13 +2,52 @@
 import copy
 import dataclasses
 import importlib
+import inspect
 import logging
-from typing import Any, Dict, Iterator, Mapping
+from typing import Any, Callable, Dict, Iterator, Mapping, Union
 
 from burr.core import serde
 
 logger = logging.getLogger(__name__)
 
+FIELD_SERIALIZATION = {}
+
+
+def register_field_serde(field_name: str, serializer: Callable, deserializer: Callable):
+    """Registers a custom serializer + deserializer for a field globally.
+
+    This is useful for really controlling how a field is serialized and deserialized for
+    tracking / persistence.
+
+    .. code-block:: python
+
+        def my_field_serializer(value: MyType, **kwargs) -> dict:
+            serde_value = _do_something_to_serialize(value)
+            return {"value": serde_value}
+
+        def my_field_deserializer(value: dict, **kwargs) -> MyType:
+            serde_value = value["value"]
+            return _do_something_to_deserialize(serde_value)
+
+        register_field_serde("my_field", my_field_serializer, my_field_deserializer)
+
+    :param field_name: The name of the field to register the serializer for.
+    :param serializer: A function that takes the field value and returns a JSON serializable object.
+    :param deserializer: A function that takes the JSON serializable object and returns the field value.
+    """
+    # assert that the serializer has **kwargs argument; it also needs to return a dict but we can't check that.
+    # def name(value: Any, **kwargs) -> dict
+    serializer_sig = inspect.signature(serializer)
+    if not any(param.name == "kwargs" for param in serializer_sig.parameters.values()):
+        raise ValueError(f"Serializer for [{field_name}] must have **kwargs argument.")
+
+    # assert that the deserializer has **kwargs argument; it also needs to return a dict but we can't check that.
+    deserializer_sig = inspect.signature(deserializer)
+    if not any(param.name == "kwargs" for param in deserializer_sig.parameters.values()):
```

Comment:
The better way to do this is to assert that the first parameter is of kind `VAR_KEYWORD`. Can also check its name (but I'm not sure that's relevant?).

```python
In [1]: def foo(**kwargs):
   ...:     pass
   ...:
In [2]: inspect.signature(foo).parameters['kwargs'].kind
Out[2]: <_ParameterKind.VAR_KEYWORD: 4>

In [13]: inspect.Parameter.VAR_KEYWORD
Out[13]: <_ParameterKind.VAR_KEYWORD: 4>
```

Reply:
Oh sure:
```python
if not any(param.kind == param.VAR_KEYWORD for param in deserializer_sig.parameters.values()):
```
checks for the right thing